src/main/scala/util/Replacement.scala

```diff
@@ -37,43 +63,147 @@ class SeqRandom(n_ways: Int) extends SeqReplacementPolicy {
+    }
+    for (i <- 1 until ways.size) {
+      cover(PopCount(ways.map(_.valid)) === i.U, s"LRU_UpdateCount$i", s"LRU Update $i simultaneous")
+    }
   }
-  def get_next_state(state: UInt, way: UInt) = {
-    var next_state = state << 1
-    var idx = UInt(1,1)
-    for (i <- log2Up(n)-1 to 0 by -1) {
-      val bit = way(i)
-      next_state = next_state.bitSet(idx, !bit)
-      idx = Cat(idx, bit)
+
+  def get_replace_way(state: UInt): UInt = {
+    val moreRecentVec = extractMRUVec(state)  // reconstruct lower triangular matrix
+    // For each way, determine if all other ways are more recent
+    val mruWayDec     = (0 until n_ways).map { i =>
+      val upperMoreRecent = (if (i == n_ways-1) true.B else moreRecentVec(i).apply(n_ways-1,i+1).andR)
+      val lowerMoreRecent = (if (i == 0)        true.B else moreRecentVec.map(e => !e(i)).reduce(_ && _))
+      upperMoreRecent && lowerMoreRecent
     }
-    next_state.extract(n-1, 1)
+    OHToUInt(mruWayDec)
   }
-  def replace = get_replace_way(state_reg)
-  def get_replace_way(state: UInt) = {
-    val shifted_state = state << 1
-    var idx = UInt(1,1)
-    for (i <- log2Up(n)-1 to 0 by -1) {
-      val in_bounds = Cat(idx, UInt(BigInt(1) << i))(log2Up(n)-1, 0) < UInt(n)
-      idx = Cat(idx, in_bounds && shifted_state(idx))
+
+  def way = get_replace_way(state_reg)
+  def miss = access(way)
+  def hit = {}
+}
+
+class PseudoLRU(n_ways: Int) extends ReplacementPolicy {
+  //example bits storage format for 4-way PLRU:
+  // [2] = ways 3-2 older than ways 1-0
+  // [1] = way 3 older than way 2
+  // [0] = way 1 older than way 0
+  def nBits = n_ways - 1
+  private val state_reg = Reg(UInt(nBits.W))
+  def state_read = WireDefault(state_reg)
+
+  def access(way: UInt) {
+    state_reg := get_next_state(state_reg, way)
+  }
+  def access(ways: Seq[Valid[UInt]]) {
+    when (ways.map(_.valid).orR) {
+      state_reg := get_next_state(state_reg, ways)
+    }
+    for (i <- 1 until ways.size) {
+      cover(PopCount(ways.map(_.valid)) === i.U, s"PLRU_UpdateCount$i", s"PLRU Update $i simultaneous")
+    }
+  }
+
+  def get_next_state(state: UInt, way: UInt, this_ways: Int): UInt = {
+    require(state.getWidth == (this_ways-1), s"wrong state bits width ${state.getWidth} for $this_ways ways")
+    require(way.getWidth == (log2Ceil(this_ways) max 1), s"wrong encoded way width ${way.getWidth} for $this_ways ways")
+    if (this_ways > 2) {
+      val half_ways: Int = 1 << (log2Ceil(this_ways) - 1)
+      if (this_ways > 3) {
+        Cat(!way(log2Ceil(this_ways)-1),
+            Mux(way(log2Ceil(this_ways)-1),
+                get_next_state(state(this_ways-3,half_ways-1), way(log2Ceil(this_ways-half_ways)-1,0), this_ways-half_ways),
+                state(this_ways-3,half_ways-1)),
+            Mux(way(log2Ceil(this_ways)-1),
+                state(half_ways-2,0),
+                get_next_state(state(half_ways-2,0), way(log2Ceil(half_ways)-1,0), half_ways)))
```

Comment:
@johningalls-sifive can you add some comments here or in another file that explains _**in detail**_ what is going on? The software-engineer in me sees all these magic -3, -2, -1 values, and the lack of comments, and although I trust you that it does the right thing, I don't know if anyone besides you can maintain this function as it currently is! :)

Reply:
I agree completely and went through the PseudoLRU code again.  Let me know if it's improved now!